### ISSUE TITLE
Possible mse_loss bug for device=mps

### ISSUE THREAD

**Issue by Wissiak** [pytorch/pytorch#121025]
### 🐛 Describe the bug

I am adapting the YOLOv8 model which uses pytorch. I want to measure the Euclidean distance between two tensors using `torch.nn.MSELoss`:
```
loss_fn = nn.MSELoss()
loss_fn(gt_corners, corners)
```
It is working fine for the training process (batch size = 16) but during the validation phase (batch size = 32) I get the following error:
```
loc("mps_subtract"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/4e1473ee-9f66-11ee-8daf-cedaeb4cabe2/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":233:0)): error: input types 'tensor<300x24xf32>' and 'tensor<300x24xf16>' are not broadcast compatible
```
And also:
```
UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
```
The error message is pretty straight forward but the weird thing is that if I debug the line of the `loss_fn` call I get `gt_corners.shape=torch.Size([150, 24])` and `corners.shape=torch.Size([150, 24])` in both the training and validation phases. So there does not seem to be the batch size to be involved anywhere in the loss function call... 
The same behavior appears when using `torch.nn.functional.mse_loss(gt_corners, corners)`.

However, if I run the code with device="cpu" or on a NVIDIA GPU with device="cuda", I don't get an error and it works just fine... This is why I think it could be a bug in the library.

You can see the code in my [forked YOLOv8 repository](https://github.com/Wissiak/ultralytics/blob/68b4f3554468398a5a54f291363934e70d9bf24f/ultralytics/utils/loss.py#L831C1-L836C55). 

### Versions

PyTorch version: 2.2.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.3.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: version 3.28.3
Libc version: N/A

Python version: 3.9.18 | packaged by conda-forge | (main, Dec 23 2023, 16:35:41)  [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-14.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.1
[pip3] torchvision==0.17.1
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.2.1                    pypi_0    pypi
[conda] torchvision               0.17.1                   pypi_0    pypi

**Comment by malfet** [comment]
Looks like your inputs are mix of fp32 and fp16

**Comment by Wissiak** [comment]
You're totally right 👀 using the same type fixed it - thank you.
